README.md
# Advanced Infinite Scroll

2023 © Bikramaditya Meher

[![Pub](https://img.shields.io/pub/v/advanced_infinite_scroll.svg)](https://pub.dartlang.org/packages/advanced_infinite_scroll) [![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://github.com/bikram0000/advanced_infinite_scroll/blob/master/LICENSE)

A versatile Flutter package for implementing advanced scrolling. Whether you're looking for infinite scrolling, responsive grid views, or customizable loaders, `AdvancedInfiniteScroll` has got you covered.

<img src="example/assets/ezgif-4-9547c1c63f.gif" width="320"> <img src="example/assets/ezgif-4-2b42dfdc9e.gif" width="320">


## Features:

- Infinite Scrolling
- Responsive Grid/List Views
- Pull-to-Refresh Capability
- Customizable Loaders For Loading More or Initial Loader
- "No Data Found" Widget Handling
- "ON ERROR" Widget Handling,
- Optimized Rendering for Visible Items
- Header/Footer widget
- empty loader size (check example)

## Usage:

To use this package, add `advanced_infinite_scroll` as a dependency in your `pubspec.yaml` file.

### Data Fetching:

You can fetch data from your network or any source. For demonstration purposes, here's a dummy data-fetching function:

```dart
Future<List<String>> onListFutureDummy(int page, int perPage, Map? params) async {
  debugPrint("ON LOAD DATA AIS :: $page");
  await Future.delayed(const Duration(seconds: 1));
  return List.generate(perPage, (index) => "PAGE :: $page ::");
  // return [];
}
```

This function simulates a network call with a delay and generates dummy data.


### Basic Setup:

Here's a simple example demonstrating the usage:

```dart
     AdvancedInfiniteScroll<String>(
        minItemWidth: 120,
        minItemsPerRow: 1,
        maxItemsPerRow: 1,
        controller: controller,
        loaderSize: 1,
        headerWidget: Container(
          height: 50,
          color: Colors.green,
          child: const Center(child: Text("Header")),
        ),
        footerWidget: Container(
          height: 50,
          color: Colors.red,
          child: const Center(child: Text("Header")),
        ),
        noDataFoundWidget: (c) {
          return TextButton(
            onPressed: () {
              c.refresh();
            },
            child: const Text("Refresh"),
          );
        },
        loadingWidget: const Center(child: CircularProgressIndicator()),
        // onFuture: onListFutureDummy,
        loadingMoreWidget: Center(
          child: Padding(
            padding: const EdgeInsets.all(10),
            child: LinearProgressIndicator(
              minHeight: 60,
              valueColor: AlwaysStoppedAnimation<Color>(Colors.grey.shade50),
            ),
          ),
        ),
        builder: (BuildContext context, listData, index) {
          return ListTile(
            title: Text(
              "${listData[index]}:: INDEX :: $index ::",
            ),
          );
        },
      ),
```

### Parameters:

Here's a brief overview of the key parameters:

- `minItemWidth`: The minimum width for an item.
- `minItemsPerRow`: The minimum number of items per row.
- `controller`: The controller associated with `AdvancedInfiniteScroll`.
- `loadingMoreWidget`: A widget to display while more items are being loaded.
- `builder`: A function that returns a list of widgets based on the provided data.


For more information and please checkout example folder.


## Contributing:

Feel free to submit issues or pull requests to enhance the package. Contributions are always welcome!

CHANGELOG.md
## 0.0.2

* Header/Footer widget added.
* Optimize builder with index
* Empty loader lines size added for both grid and list view

## 0.0.1

* Initial version.


test/advance_infinite_scroll_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:advanced_infinite_scroll/advanced_infinite_scroll.dart';

void main() {
  test('adds one to input values', () {
    // final calculator = Calculator();
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
  });
}


lib/advance_infinite_scroll.dart
library advance_infinite_scroll;

import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:quiver/iterables.dart';
import 'package:responsive_grid_list/responsive_grid_list.dart';
import 'package:responsive_grid_list/src/extensions/list_extensions.dart';

class AdvanceInfiniteScrollController<T> {
  /// when rendering it will return numbers of item in a row.
  final Function(int)? onItemCount;
  /// when rendering it will return item's width
  final Function(double)? onItemWidth;
  /// it is for getting numbers of items per page from network by default
  /// it will 10.
  final int perPage;
  /// to get the state of [AdvanceInfiniteScroll]
  AdvanceInfiniteScrollState<T>? _myWidgetState;


  /// Callback function on pull up to load more data and refresh
  final Future<List<T>?> Function(int page, int perPage, Map? params) onFuture;

  void _bind(AdvanceInfiniteScrollState<T> state) {
    _myWidgetState = state;
  }

  AdvanceInfiniteScrollController(
      {this.onItemCount,
      this.onItemWidth,
      required this.onFuture,
      this.perPage = 10});

  /// TO refresh the network call.
  Future<List<T>?> refresh({
    Map? params,
  }) async {
    return await _myWidgetState?.loadFutureList(params: params);
  }
}


class AdvanceInfiniteScroll<T> extends StatefulWidget {
  /// Whether need pull refresh or not.
  final bool pullRefresh;
  /// Builder for getting list of widget. [listData] it will return the same getting from network.
  final List<Widget> Function(BuildContext context, List<T> listData) builder;


  final Widget? loadingMoreWidget;
  final Widget? loadingWidget;


  final Widget Function(AdvanceInfiniteScrollController<T> controller)?
      noDataFoundWidget;
  final Widget Function(AdvanceInfiniteScrollController<T> controller)?
      errorWidget;

  ///
  /// Object that can be set if any of the [ListView.builder] options
  /// need to be overridden. The [ResponsiveGridList] defines the builder
  /// and item count. All other options are optional and can be set through
  /// this object.
  ///
  final ListViewBuilderOptions? listViewBuilderOptions;

  ///
  /// The minimum item width of each individual item in the list. Can be smaller
  /// if the viewport constraints are smaller.
  ///
  final double minItemWidth;

  ///
  /// Minimum items to show per row. If this is set to a value higher than 1,
  /// this takes precedence over [minItemWidth] and allows items to be smaller
  /// than [minItemWidth] to fit at least [minItemsPerRow] items.
  ///
  final int minItemsPerRow;

  ///
  /// Maximum items to show per row. By default the package shows all items that
  /// fit into the available space according to [minItemWidth].
  ///
  /// Note that this should only be used when limiting items on large screens
  /// since it will stretch [maxItemsPerRow] items across the whole width
  /// when maximum is reached. This can result in a large difference to
  /// [minItemWidth].
  ///
  final int? maxItemsPerRow;

  ///
  /// The horizontal spacing between the items in the grid.
  ///
  final double horizontalGridSpacing;

  ///
  /// The vertical spacing between the items in the grid.
  ///
  final double verticalGridSpacing;

  ///
  /// The horizontal spacing around the grid.
  ///
  final double? horizontalGridMargin;

  ///
  /// The vertical spacing around the grid.
  ///
  final double? verticalGridMargin;

  ///
  /// [MainAxisAlignment] of each row in the grid list.
  ///
  final MainAxisAlignment rowMainAxisAlignment; // coverage:ignore-end

  ///
  ///[AdvanceInfiniteScrollController] it can help to refresh the list manually,
  ///
  final AdvanceInfiniteScrollController<T> controller;

  const AdvanceInfiniteScroll({
    super.key,
    this.listViewBuilderOptions,
    required this.minItemWidth,
    required this.minItemsPerRow,
    this.horizontalGridSpacing = 1,
    this.verticalGridSpacing = 1,
    this.rowMainAxisAlignment = MainAxisAlignment.center,
    this.maxItemsPerRow,
    this.horizontalGridMargin,
    this.verticalGridMargin,

    ///related to load more..
    this.pullRefresh = true,
    required this.builder,

    ///loader..
    this.loadingMoreWidget = const Center(child: CircularProgressIndicator()),
    this.loadingWidget,
    this.noDataFoundWidget,
    this.errorWidget,
    required this.controller,
  })  : assert(
          // coverage:ignore-start
          minItemWidth > 0,
          'minItemWidth has to be > 0. It instead was set to $minItemWidth',
        ),
        assert(
          minItemsPerRow > 0,
          'minItemsPerRow has to be > 0. It instead was set to $minItemsPerRow',
        ),
        assert(
          maxItemsPerRow == null || maxItemsPerRow >= minItemsPerRow,
          'maxItemsPerRow can only be null or >= minItemsPerRow '
          '($minItemsPerRow). It instead was set to $maxItemsPerRow',
        );

  ///
  /// Method to generate a list of [ResponsiveGridRow]'s with spacing in between
  /// them.
  ///
  /// [maxWidth] is the maximum width of the current layout.
  ///
  List<Widget> getResponsiveGridListItems(
      double maxWidth, List<Widget> children) {
    // Start with the minimum allowed number of items per row.
    var itemsPerRow = minItemsPerRow;

    // Calculate the current width according to the items per row
    var currentWidth =
        itemsPerRow * minItemWidth + (itemsPerRow - 1) * horizontalGridSpacing;

    // Add outer margin (vertical) if set
    if (horizontalGridMargin != null) {
      currentWidth += 2 * horizontalGridMargin!;
    }

    // While another pair of spacing + minItemWidth fits the row, add it to
    // the variables. Only add items while maxItemsPerRow is not reached.
    while (currentWidth < maxWidth &&
        (maxItemsPerRow == null || itemsPerRow < maxItemsPerRow!)) {
      if (currentWidth + (minItemWidth + horizontalGridSpacing) <= maxWidth) {
        // If another spacing + item fits in the row, add one item to the row
        // and update the currentWidth
        currentWidth += minItemWidth + horizontalGridSpacing;
        itemsPerRow++;
      } else {
        // If no other item + spacer fits into the row, break
        break;
      }
    }

    // Calculate the spacers per row (they are only in between the items, not
    // at the edges)
    final spacePerRow = itemsPerRow - 1;

    // Calculate the itemWidth that results from the maxWidth and number of
    // spacers and outer margin (horizontal)
    final itemWidth = (maxWidth -
            (spacePerRow * horizontalGridSpacing) -
            (2 * (horizontalGridMargin ?? 0))) /
        itemsPerRow;

    // Partition the items into groups of itemsPerRow length and map them
    // to ResponsiveGridRow's
    final items = partition(children, itemsPerRow)
        .map<Widget>(
          (e) => ResponsiveGridRow(
            rowItems: e,
            spacing: horizontalGridSpacing,
            horizontalGridMargin: horizontalGridMargin,
            itemWidth: itemWidth,
            rowMainAxisAlignment: rowMainAxisAlignment,
          ),
        )
        .toList();

    // Join the rows width spacing in between them (vertical)
    final responsiveGridListItems =
        items.genericJoin(SizedBox(height: verticalGridSpacing));

    // Add outer margin (vertical) if set
    if (verticalGridMargin != null) {
      return [
        SizedBox(height: verticalGridMargin),
        ...responsiveGridListItems,
        SizedBox(height: verticalGridMargin)
      ];
    }

    return responsiveGridListItems;
  }

  @override
  AdvanceInfiniteScrollState<T> createState() =>
      AdvanceInfiniteScrollState<T>();
}

class AdvanceInfiniteScrollState<T> extends State<AdvanceInfiniteScroll<T>> {
  List<T>? futureList;
  int page = 1;
  bool? isLastPage;
  bool loadingFuture = false;

  @override
  void initState() {
    super.initState();
    loadFutureList();
  }

  @override
  Widget build(BuildContext context) {
    widget.controller._bind(this);
    if ((futureList == null || (futureList as List).isEmpty)) {
      if (loadingFuture) {
        return (widget.loadingWidget ??
            widget.loadingMoreWidget ??
            const Center(child: CircularProgressIndicator()));
      } else if (!loadingFuture && futureList != null) {
        ///its means it is empty
        return (widget.noDataFoundWidget != null
            ? widget.noDataFoundWidget!(widget.controller)
            : const Center(child: Text("NO DATA FOUND")));
      } else {
        return (widget.errorWidget != null
            ? widget.errorWidget!(widget.controller)
            : const Center(child: Text("SOMETHING WENT WRONG !!")));
      }
    }
    return LayoutBuilder(
      builder: (BuildContext context, BoxConstraints constraints) {
        // Get the grid list items
        final items = widget.getResponsiveGridListItems(
          constraints.maxWidth -
              (widget.listViewBuilderOptions?.padding?.horizontal ?? 0),
          widget.builder(
            context,
            futureList!,
          ),
        );

        if (widget.controller.onItemCount != null) {
          if (items.first is ResponsiveGridRow) {
            if (widget.controller.onItemCount != null) {
              widget.controller.onItemCount!(
                  (items.first as ResponsiveGridRow).rowItems.length);
            }
            if (widget.controller.onItemWidth != null) {
              widget.controller
                  .onItemWidth!((items.first as ResponsiveGridRow).itemWidth);
            }
          }
        }
        Widget child = ListView.builder(
          itemCount: items.length + (widget.loadingMoreWidget != null ? 1 : 0),
          itemBuilder: (BuildContext context, int index) {
            if (index >= items.length) {
              if (isLastPage ?? true) {
                return const SizedBox();
              } else {
                loadFutureList(loadMore: true);
                return widget.loadingMoreWidget;
              }
            } else {
              return items[index];
            }
          },
          scrollDirection:
              widget.listViewBuilderOptions?.scrollDirection ?? Axis.vertical,
          reverse: widget.listViewBuilderOptions?.reverse ?? false,
          controller: widget.listViewBuilderOptions?.controller,
          primary: widget.listViewBuilderOptions?.primary,
          physics: widget.listViewBuilderOptions?.physics,
          shrinkWrap: widget.listViewBuilderOptions?.shrinkWrap ?? false,
          padding: widget.listViewBuilderOptions?.padding,
          itemExtent: widget.listViewBuilderOptions?.itemExtent,
          prototypeItem: widget.listViewBuilderOptions?.prototypeItem,
          findChildIndexCallback:
              widget.listViewBuilderOptions?.findChildIndexCallback,
          addAutomaticKeepAlives:
              widget.listViewBuilderOptions?.addAutomaticKeepAlives ?? true,
          addRepaintBoundaries:
              widget.listViewBuilderOptions?.addRepaintBoundaries ?? true,
          addSemanticIndexes:
              widget.listViewBuilderOptions?.addSemanticIndexes ?? true,
          cacheExtent: widget.listViewBuilderOptions?.cacheExtent,
          semanticChildCount: widget.listViewBuilderOptions?.semanticChildCount,
          dragStartBehavior: widget.listViewBuilderOptions?.dragStartBehavior ??
              DragStartBehavior.start,
          keyboardDismissBehavior:
              widget.listViewBuilderOptions?.keyboardDismissBehavior ??
                  ScrollViewKeyboardDismissBehavior.manual,
          restorationId: widget.listViewBuilderOptions?.restorationId,
          clipBehavior:
              widget.listViewBuilderOptions?.clipBehavior ?? Clip.hardEdge,
        );
        return !widget.pullRefresh
            ? child
            : ScrollConfiguration(
                behavior: ScrollConfiguration.of(context).copyWith(
                  dragDevices: {
                    PointerDeviceKind.touch,
                    PointerDeviceKind.mouse,
                    PointerDeviceKind.trackpad,
                    PointerDeviceKind.stylus,
                    PointerDeviceKind.invertedStylus,
                  },
                ),
                child: RefreshIndicator(
                  color: Theme.of(context).primaryColor,
                  onRefresh: () async {
                    page = 1;
                    await loadFutureList();
                  },
                  child: child,
                ),
              );
      },
    );
  }

  Future<List<T>?> loadFutureList({bool loadMore = false, Map? params}) async {
    if (loadingFuture) {
      return null;
    }
    if (loadMore) {
      page++;
    } else {
      page = 1;
    }
    if (page == 1) {
      futureList = null;
    }
    loadingFuture = true;
    update();
    try {
      List<T>? futureList2 = await widget.controller
          .onFuture(page, widget.controller.perPage, params);
      if (futureList2 != null) {
        if (page == 1) {
          futureList = futureList2;
        } else {
          futureList?.addAll(futureList2);
        }

        if ((futureList2).length == widget.controller.perPage) {
          isLastPage = false;
        } else {
          isLastPage = true;
        }
      }
    } catch (e) {
      debugPrint("ERROR AIS ::  $e");
    }
    loadingFuture = false;
    update();
    return futureList;
  }

  void update() {
    WidgetsBinding.instance.addPostFrameCallback((_) {
      if (mounted) {
        setState(() {});
      }
    });
  }
}


example/lib/main.dart
import 'package:advanced_infinite_scroll/advanced_infinite_scroll.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // TRY THIS: Try running your application with "flutter run". You'll see
        // the application has a blue toolbar. Then, without quitting the app,
        // try changing the seedColor in the colorScheme below to Colors.green
        // and then invoke "hot reload" (save your changes or press the "hot
        // reload" button in a Flutter-supported IDE, or press "r" if you used
        // the command line to start the app).
        //
        // Notice that the counter didn't reset back to zero; the application
        // state is not lost during the reload. To reset the state, use hot
        // restart instead.
        //
        // This works for code too, not just values: Most code changes can be
        // tested with just a hot reload.
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  late AdvancedInfiniteScrollController<String> controller;

  @override
  void initState() {
    super.initState();
    controller = AdvancedInfiniteScrollController(
      onFuture: onListFutureDummy,
      perPage: 14,
    );
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // TRY THIS: Try changing the color here to a specific color (to
        // Colors.amber, perhaps?) and trigger a hot reload to see the AppBar
        // change color while the other colors stay the same.
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: AdvancedInfiniteScroll<String>(
        minItemWidth: 120,
        minItemsPerRow: 1,
        maxItemsPerRow: 1,
        controller: controller,
        loaderSize: 1,
        headerWidget: Container(
          height: 50,
          color: Colors.green,
          child: const Center(child: Text("Header")),
        ),
        footerWidget: Container(
          height: 50,
          color: Colors.red,
          child: const Center(child: Text("Header")),
        ),
        noDataFoundWidget: (c) {
          return TextButton(
            onPressed: () {
              c.refresh();
            },
            child: const Text("Refresh"),
          );
        },
        loadingWidget: const Center(child: CircularProgressIndicator()),
        // onFuture: onListFutureDummy,
        loadingMoreWidget: Center(
          child: Padding(
            padding: const EdgeInsets.all(10),
            child: LinearProgressIndicator(
              minHeight: 60,
              valueColor: AlwaysStoppedAnimation<Color>(Colors.grey.shade50),
            ),
          ),
        ),
        builder: (BuildContext context, listData, index) {
          return ListTile(
            title: Text(
              "${listData[index]}:: INDEX :: $index ::",
            ),
          );
        },
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          controller.refresh(params: {"Sadf": "afdas"});
        },
        child: const Icon(Icons.refresh),
      ),
    );
  }

  Future<List<String>> onListFutureDummy(
      int page, int perPage, Map? params) async {
    debugPrint("ON LOAD DATA AIS :: $page");
    await Future.delayed(const Duration(seconds: 1));
    return List.generate(perPage, (index) => "PAGE :: $page ::");
    // return [];
  }
}
